design/ceph-openshift-fencing-mitigation.md

```diff
@@ -0,0 +1,70 @@
+
+
+# Problem Statement
+We need to ensure that only one node can be fenced at a time and that Ceph is fully recovered (has PGs clean) before any fencing is initiated. The available pattern for limiting fencing is the MachineDisruptionBudget which allows us to specify maxUnavailable. However, this won’t be sufficient to ensure that Ceph has recovered before fencing is initiated as MachineHealthCheck does not check anything other than the node state.
+
+Therefore, we will control how many nodes match the MDB by dynamically adding and removing labels as well as dynamically updating the MDB. By manipulating the MDB into a state where desiredHealthy > currentHealthy, we can disable fencing on the nodes the MDB points to.
+
+# Design:
+
+We will implement two controllers `machinedisruptionbudget-controller` and the `machine-controller` to be implemented through the controller pattern describere [here](https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg#hdr-Controller_Writing_Tips). Each controller watches a set of object kinds and reconciles one.
+
+## machinedisruptionbudget-controller
+This controller watches ceph PGs and CephClusters. We will ensure the reconciler is enqueued every 60s. It ensures that each CephCluster has a MDB created, and the MDB's value of maxUnvailable reflects the health of the Ceph Cluster's PGs.
+If all PGs are clean, maxUnavailable = 1.
+else, maxUnavailable = 0.
+
+We can share a ceph health cache with the other controller-runtime reconcilers that have to watch the PG "cleanliness".
+
+The MDB will target `Machine`s selected by a label maintainted by the `machine-controller`. The label is `fencegroup.rook.io/<cluster-name>`. 
+
+## machine-controller
+This controller watches OSDs and `Machine`s. It ensures that each `Machine` with OSDs from a `CephCluster` have the label `fencegroup.rook.io/<cluster-name>`, and those that do not have running OSDs do not have label.
+
+This will ensure that no `Machine` without running OSDs will be protected by the MDB.
+
+## Assumptions:
+- We assume that the controllers will be able to reconcile multiple times in < 5 minutes as we know that fencing will happen only after a configurable timeout. The default timeout is 5 minutes.
+  This is important because the MDB must be reconciled based on an accurate ceph health state in that time.
+
+
+## Example Flows:
+
+**Node needs to be fenced, the OSDs on the node are down too**
+
+ - Node has NotReady condition.
+ - Some Ceph PGs are not active+clean.
+ - machinedisruptionbudget-controller sets maxUnavailable to 0 on the MachineDisruptionBudget.
+ - MachineHealthCheck sees NotReady and attempts to fence after 5 minutes, but can't due to MDB
+ - machine-controller notices all OSDs on the affected node are down and removes the node from the MDB.
```

Comment:
@rohantmp Can you elaborate in the determination of OSDs being down? Is it pod health, osd daemon health, osd state down or down+out for example. Do all the OSDs on an instance have the same node details in the osd metadata to be able to make the determination that all osds on a node are down - and if so, does this node name relate 100% to the k8s node name.

Reply:
This is something I think we'll have to fiddle with during the implementation. I think OSD state DOWN is the best candidate.

Reply:
+1 osd state down seems like a good option. Would be good to clarify the metadata questions too.

Reply:
The OSD metadata might not have the information about which node it's on, but we can make that connection by following the labeled pods with their osd id. There is sufficient metadata to do that.

Reply:
@pcuzner does that address the metadata concern?

Reply:
So you're using out-of-band communication to know that the OSDs are gone, and then removing the `fencegroup.rook.io/<cluster-name>` label?